### ISSUE TITLE
Encoding detection logic?

### ISSUE THREAD

**Issue by OpportunityLiu** [zzzprojects/html-agility-pack#213]
I've tried some html document with gbk encoding and `<meta http-equiv="Content-Type" content="text/html; charset=gbk" />`.

It seems that html-agility-pack can detect the right encoding, since `DeclaredEncoding` has been set to correct value after `HtmlDocument.Load()`, but the content of document is wrong. I have to load the document again with the right encoding.

Here's a project for reproduction:
https://github.com/OpportunityLiu/HtmlAgilityPack-EncodingDetection

**Comment by JonathanMagnan** [comment]
Hello @OpportunityLiu ,

Unfortunately, the encoding is found via the charset `gbk` which mean the text must already be read a first time to detect the right encoding.

We tried with the method `doc2.DetectEncoding(new MemoryStream(data));` but it seem to return the wrong encoding. I believe that's due because the HTML is not read to be faster, so the charset is not used.

At this moment, we have two possible solutions that we could do:
- Add an option ReloadWithEncoding (or something similar)
- Fix the DetectEncoding to also read the current charset, but that means the document will need to be loaded.

Let me know your thought about the best solution.

Best Regards,

Jonathan

**Comment by OpportunityLiu** [comment]
`DetectEncoding(Stream)` use encoding detection by `StreamReader` (see: https://github.com/zzzprojects/html-agility-pack/blob/3db9cd80976a2c888240b51ae2c8872fea7bcddc/src/HtmlAgilityPack/HtmlDocument.cs#L474-L481),
which only detects Unicode encodings (see: https://github.com/dotnet/coreclr/blob/556a5fa3e898bbaad3516b55875074a5c265327a/src/System.Private.CoreLib/shared/System/IO/StreamReader.cs#L488). 


Is it reliable enough to read n chars in the beginning of the stream to detect encoding? I tried to parse following content and encoding is successfully detected.
```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=gbk" />
    <titl
```

**Comment by JonathanMagnan** [comment]
> Is it reliable enough to read first n chars of the stream to detect the encoding

I would say no. charset is normally the first meta but nothing force to be the first.

If we read the first 200 characters, then a page that's modified with some additional heading meta could lead eventually to an error when it was working the previous day.

--- 
However, we can always read the whole file as a string and use `regex` or simply some `IndexOf` to find as fast as possible the charset instead of parsing all the text with HAP. But again, that could lead to some error when the text is in a comment.

So maybe only reading the meta section with HAP and stop as soon as we reach the body tag?

Let me know if you have a better idea

**Comment by OpportunityLiu** [comment]
> So maybe only reading the meta section with HAP and stop as soon as we reach the body tag?
Agreed, that might be the only method to ensure correctness.

**Comment by JonathanMagnan** [comment]
Great ;)

Unless we think about a better idea, I will have one of my developer using this logic to improve the DetectEncoding method.

**Comment by JonathanMagnan** [comment]
Hello @OpportunityLiu ,

The v1.8.5 has been released.

We added a new overload `DetectEncoding(Stream, true)` that will check in the HTML instead of using the stream encoding when the second parameter is true.

It will require you to call the `DetectEncoding` method but at least, it will look less weird then reading twice the stream.

Let me know if that fixes your issue.

Best Regards,

Jonathan
